# Limpieza de datos

Para nuestro análisis exploratorio necesitaremos los datos de las siguientes columnas (* datos imprescindibles):

<u>Planeta</u>
- pl_name  
- (*)pl_rade: radio en radios terrestres (R<sub>Terrestre</sub>)
- (*)pl_bmasse: masa en masas terrestres
- (*)pl_eqt: temperatura de equilibrio
- pl_orbper: periodo orbital (días)
- pl_orbsmax: semieje mayor (distancia a la estrella)
- (*)pl_insol: insolación recibida

<u>Estrella</u>  
- hostname: nombre de la estrella
- st_teff: temperatura efectiva
- st_rad: radio
- (*)st_mass: masa
- st_met: metalicidad
- st_spectype: tipo de estrella/tipo espectral (G, K, M…)  
>>Principales Tipos Espectrales (de más caliente a más frío):  
>>>O (Azules): (>30 000 K) Muy masivas, raras y brillantes.  
>>>B (Azules/Blancas): (10 000 - 30 000 K) Brillantes, helio ionizado.  
>>>A (Blancas): (7 500 - 10 000 K) Líneas de hidrógeno fuertes.  
>>>F (Blanco-Amarillas): (6 000 - 7 500 K)   
>>>G (Amarillas): (5 200 - 6 000 K) Estrellas como el Sol.  
>>>K (Naranjas): (3 500 - 5 200 K) Más frías y tenues que el Sol.  
>>>M (Rojas): (2 500 - 3 500 K) Las más abundantes, muy frías.

>>Clases de Luminosidad (Tamaño):  
>>>I: Supergigantes  
>>>II: Gigantes luminosas  
>>>III: Gigantes  
>>>IV: Subgigantes  
>>>V: Secuencia Principal (enanas)  

- st_age: edad estimada
- st_rotp: periodo de rotación

<u>Sistema</u>  
- (*)sy_dist: distancia en parsecs
- sy_pnum: número de planetas en el sistema
- (*)discoverymethod: método de detección
- disc_year: año descubierto


In [1]:
import pandas as pd
import numpy as np

In [2]:
PSCP_api = pd.read_csv('../data/pscomppars_API.csv') # Ruta lógica desde donde se encuentra el dataset

C:\Users\urkow\AppData\Local\Temp\ipykernel_25952\3554068022.py:1: DtypeWarning: Columns (0: pl_orbinclstr, 1: pl_orbincl_reflink, 2: pl_eqtstr, 3: pl_eqt_reflink, 4: pl_occdepstr, 5: pl_occdep_reflink, 6: pl_insolstr, 7: pl_insol_reflink, 8: sy_umagstr, 9: sy_rmagstr, 10: sy_imagstr, 11: sy_zmagstr, 12: sy_gmagstr, 13: st_rotpstr, 14: pl_projobliqstr, 15: pl_densstr, 16: pl_trandepstr, 17: pl_trandep_reflink, 18: pl_trandurstr, 19: pl_trandur_reflink, 20: pl_radjstr, 21: pl_radestr, 22: pl_ratrorstr, 23: pl_ratror_reflink, 24: pl_trueobliqstr, 25: pl_trueobliq_reflink, 26: sy_icmagstr, 27: pl_ratdorstr, 28: pl_ratdor_reflink, 29: pl_impparstr, 30: pl_imppar_reflink, 31: pl_cmassjstr, 32: pl_cmassj_reflink, 33: pl_cmassestr, 34: pl_cmasse_reflink, 35: pl_projobliq_reflink) have mixed types. Specify dtype option on import or set low_memory=False.
  PSCP_api = pd.read_csv('../data/pscomppars_API.csv') # Ruta lógica desde donde se encuentra el dataset


Creamos nuevo df con las columnas que necesitamos

In [3]:
exoplanets = PSCP_api[['rowid','pl_name','pl_rade','pl_bmasse','pl_eqt','pl_orbper','pl_orbsmax','pl_insol','hostname','st_teff','st_rad','st_mass','st_met','st_spectype','st_age','st_rotp','sy_dist','sy_pnum','discoverymethod','disc_year']]

In [4]:
exoplanets

,rowid,pl_name,pl_rade,pl_bmasse,pl_eqt,pl_orbper,pl_orbsmax,pl_insol,hostname,st_teff,st_rad,st_mass,st_met,st_spectype,st_age,st_rotp,sy_dist,sy_pnum,discoverymethod,disc_year
0,1,11 Com b,12.2,4914.898486,NaN,323.210000,1.178000,NaN,11 Com,4874.00,13.76,2.09,-0.2600,G8 III,NaN,NaN,93.1846,1,Radial Velocity,2007.0
1,2,11 UMi b,12.3,4684.814200,NaN,516.219970,1.530000,NaN,11 UMi,4213.00,29.79,2.78,-0.0200,K4 III,1.56,NaN,125.3210,1,Radial Velocity,2009.0
2,3,14 And b,13.1,1131.151301,NaN,186.760000,0.775000,NaN,14 And,4888.00,11.55,1.78,-0.2100,K0 III,4.50,NaN,75.4392,1,Radial Velocity,2008.0
3,4,14 Her b,12.5,2828.672822,NaN,1766.410000,2.839000,NaN,14 Her,5338.00,0.93,0.97,0.4300,K0,4.60,NaN,17.9323,2,Radial Velocity,2002.0
4,5,16 Cyg B b,13.5,565.737400,NaN,798.500000,1.660000,NaN,16 Cyg B,5750.00,1.13,1.08,0.0600,G3 V,7.40,NaN,21.1397,1,Radial Velocity,1996.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6148,6149,ups And b,14.0,218.531000,NaN,4.617033,0.059222,NaN,ups And,6156.77,1.56,1.30,0.1014,F8 V,5.00,12.0,13.4054,3,Radial Velocity,1996.0
6149,6150,ups And c,12.3,4443.241130,NaN,241.258000,0.827774,NaN,ups And,6156.77,1.56,1.30,0.1014,F8 V,5.00,12.0,13.4054,3,Radial Velocity,1999.0
6150,6151,ups And d,12.5,3257.741172,NaN,1276.460000,2.513290,NaN,ups And,6156.77,1.56,1.30,0.1014,F8 V,5.00,12.0,13.4054,3,Radial Velocity,1999.0
6151,6152,ups Leo b,14.2,162.092488,NaN,385.200000,1.180000,NaN,ups Leo,4836.00,11.22,1.48,-0.2000,G9 III,NaN,NaN,52.5973,1,Radial Velocity,2021.0


In [5]:
exoplanets.info()

<class 'pandas.DataFrame'>
RangeIndex: 6153 entries, 0 to 6152
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   rowid            6153 non-null   int64  
 1   pl_name          6153 non-null   str    
 2   pl_rade          6103 non-null   float64
 3   pl_bmasse        6122 non-null   float64
 4   pl_eqt           4590 non-null   float64
 5   pl_orbper        5819 non-null   float64
 6   pl_orbsmax       5837 non-null   float64
 7   pl_insol         4313 non-null   float64
 8   hostname         6153 non-null   str    
 9   st_teff          5863 non-null   float64
 10  st_rad           5839 non-null   float64
 11  st_mass          6145 non-null   float64
 12  st_met           5603 non-null   float64
 13  st_spectype      2308 non-null   str    
 14  st_age           4843 non-null   float64
 15  st_rotp          890 non-null    float64
 16  sy_dist          6126 non-null   float64
 17  sy_pnum          6153 non

MISSINGS

Tratamiento de missings, vamos a normalizar todos los valores nulos para un mejor manejo de datos.

In [6]:
# Normalizamos valores vacíos y nulos para posterior filtrado
exoplanets = exoplanets.replace(['', ' ', 'NaN', 'nan', 'None', 'none', pd.NA], np.nan)

No hace falta este siguiente paso, ya que al sacar la información vemos que las columnas numéricas tienen el tipo correcto, lo comentamos. Aunque debemos tener en cuenta por si se da el caso en un futuro.

In [7]:
# No hace falta este paso, al sacar la información vemos que las columnas numéricas tienen el tipo correcto, lo comentamos
# # Convertimos columnas numéricas por si vienen como strings y no pegarnos un susto después
# cols_num = ['pl_rade','pl_bmasse','pl_eqt','pl_insol','pl_orbsmax','st_teff','st_mass','st_rad','sy_dist']
# for col in cols_num:
#     exoplanets[c] = pd.to_numeric(exoplanets[c], errors='coerce')

# # Filtrado base: quedarnos con filas que tengan al menos radio,  masa y distancia
# exoplanets = exoplanets.dropna(subset=['pl_rade','sy_dist'], how='any')


El año de descubrimiento tiene el tipo float porque hay un planeta cuyo año desconocemos. Dado que no es un dato esencial para nuestro análisis, buscamos cuál es el año más común para el método de descubrimiento para darle ese valor.

In [8]:
exoplanets[exoplanets['disc_year'].isna()]

,rowid,pl_name,pl_rade,pl_bmasse,pl_eqt,pl_orbper,pl_orbsmax,pl_insol,hostname,st_teff,st_rad,st_mass,st_met,st_spectype,st_age,st_rotp,sy_dist,sy_pnum,discoverymethod,disc_year
5733,5734,TOI-5789 c,2.86,5.0,718.0,12.927748,0.1009,8.7194,TOI-5789,5185.0,0.833,0.821,-0.12,K1 V,9.4,NaN,20.4581,4,Transit,NaN


In [9]:
# Máscara de años que faltan
falta = exoplanets['disc_year'].isna()

# Método de descubrimiento del planeta al que le falta 
metodo = exoplanets['discoverymethod'][falta].iloc[0]

# Sustituimos el valor
exoplanets.loc[falta, 'disc_year'] = exoplanets.loc[exoplanets['discoverymethod'] == metodo, 'disc_year'].mode().iloc[0]

# Cambiamos a tipo int el año de descubrimiento
exoplanets['disc_year'] = exoplanets['disc_year'].astype(int)

Comprobamos que todos los años están informados y son de tipo int.

In [10]:
exoplanets.info()

<class 'pandas.DataFrame'>
RangeIndex: 6153 entries, 0 to 6152
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   rowid            6153 non-null   int64  
 1   pl_name          6153 non-null   str    
 2   pl_rade          6103 non-null   float64
 3   pl_bmasse        6122 non-null   float64
 4   pl_eqt           4590 non-null   float64
 5   pl_orbper        5819 non-null   float64
 6   pl_orbsmax       5837 non-null   float64
 7   pl_insol         4313 non-null   float64
 8   hostname         6153 non-null   str    
 9   st_teff          5863 non-null   float64
 10  st_rad           5839 non-null   float64
 11  st_mass          6145 non-null   float64
 12  st_met           5603 non-null   float64
 13  st_spectype      2308 non-null   str    
 14  st_age           4843 non-null   float64
 15  st_rotp          890 non-null    float64
 16  sy_dist          6126 non-null   float64
 17  sy_pnum          6153 non

Para nuestro análisis es fundamental que dispongamos de una serie de datos con los datos imprescindibles que mencionamos al comienzo

In [11]:
# Filtrado base: quedarnos con filas que tengan al menos radio, masa y distancia
exoplanets = exoplanets.dropna(subset=['pl_rade','pl_bmasse','sy_dist'], how='any')

In [12]:
exoplanets.shape

(6053, 20)

In [13]:
exoplanets.info()

<class 'pandas.DataFrame'>
Index: 6053 entries, 0 to 6152
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   rowid            6053 non-null   int64  
 1   pl_name          6053 non-null   str    
 2   pl_rade          6053 non-null   float64
 3   pl_bmasse        6053 non-null   float64
 4   pl_eqt           4534 non-null   float64
 5   pl_orbper        5728 non-null   float64
 6   pl_orbsmax       5746 non-null   float64
 7   pl_insol         4267 non-null   float64
 8   hostname         6053 non-null   str    
 9   st_teff          5772 non-null   float64
 10  st_rad           5751 non-null   float64
 11  st_mass          6048 non-null   float64
 12  st_met           5519 non-null   float64
 13  st_spectype      2265 non-null   str    
 14  st_age           4770 non-null   float64
 15  st_rotp          877 non-null    float64
 16  sy_dist          6053 non-null   float64
 17  sy_pnum          6053 non-null

Una vez limpios los datos nos quedamos con 6053 planetas.

No todos los planetas tienen semieje mayor o radio orbital promedio 'pl_orbsmax' informado. Vamos a calcularlo usando la tercera ley de Kepler.

In [14]:
# 1. Calcular periodo en años
exoplanets['P_yr'] = exoplanets['pl_orbper'] / 365.25

# 2. Calcular semieje mayor con Kepler SOLO donde falte pl_orbsmax
faltan = exoplanets['pl_orbsmax'].isna() & exoplanets['P_yr'].notna() & exoplanets['st_mass'].notna()

exoplanets.loc[faltan, 'pl_orbsmax'] = (
    (exoplanets.loc[faltan, 'P_yr']**2 * exoplanets.loc[faltan, 'st_mass'])**(1/3)
)

# 3. Borrar columna auxiliar
exoplanets.drop(columns=['P_yr'], inplace=True)

Comprobamos que la mayoría se ha informado.

In [15]:
exoplanets.info()

<class 'pandas.DataFrame'>
Index: 6053 entries, 0 to 6152
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   rowid            6053 non-null   int64  
 1   pl_name          6053 non-null   str    
 2   pl_rade          6053 non-null   float64
 3   pl_bmasse        6053 non-null   float64
 4   pl_eqt           4534 non-null   float64
 5   pl_orbper        5728 non-null   float64
 6   pl_orbsmax       6046 non-null   float64
 7   pl_insol         4267 non-null   float64
 8   hostname         6053 non-null   str    
 9   st_teff          5772 non-null   float64
 10  st_rad           5751 non-null   float64
 11  st_mass          6048 non-null   float64
 12  st_met           5519 non-null   float64
 13  st_spectype      2265 non-null   str    
 14  st_age           4770 non-null   float64
 15  st_rotp          877 non-null    float64
 16  sy_dist          6053 non-null   float64
 17  sy_pnum          6053 non-null

In [16]:
exoplanets[exoplanets['pl_orbsmax'].isna()]

,rowid,pl_name,pl_rade,pl_bmasse,pl_eqt,pl_orbper,pl_orbsmax,pl_insol,hostname,st_teff,st_rad,st_mass,st_met,st_spectype,st_age,st_rotp,sy_dist,sy_pnum,discoverymethod,disc_year
2007,2008,KMT-2019-BLG-1953L b,14.10,200.000000,NaN,NaN,NaN,NaN,KMT-2019-BLG-1953L,NaN,NaN,0.310,NaN,NaN,NaN,NaN,7040.000000,1,Microlensing,2020
2018,2019,KMT-2021-BLG-0322L b,12.70,2034.101805,NaN,NaN,NaN,NaN,KMT-2021-BLG-0322L,NaN,NaN,0.620,NaN,NaN,NaN,NaN,6600.000000,1,Microlensing,2021
2069,2070,KMT-2024-BLG-0404L c,4.33,17.300000,NaN,NaN,NaN,NaN,KMT-2024-BLG-0404L,NaN,NaN,0.090,NaN,M,NaN,NaN,7210.000000,2,Microlensing,2025
2071,2072,KMT-2024-BLG-1044L b,2.70,7.750000,NaN,NaN,NaN,NaN,KMT-2024-BLG-1044L,NaN,NaN,0.069,NaN,NaN,NaN,NaN,7140.000000,1,Microlensing,2024
4950,4951,MOA-2019-BLG-008L b,12.20,5720.911326,NaN,NaN,NaN,NaN,MOA-2019-BLG-008L,NaN,NaN,0.900,NaN,NaN,NaN,NaN,2700.000000,1,Microlensing,2022
5039,5040,OGLE-2015-BLG-1609L b,10.40,76.278818,NaN,NaN,NaN,NaN,OGLE-2015-BLG-1609L,NaN,NaN,0.170,NaN,NaN,NaN,NaN,3300.000000,1,Microlensing,2025
5155,5156,SR 12 AB c,12.30,4131.620000,NaN,NaN,NaN,NaN,SR 12 AB,3828.93,1.103075,NaN,NaN,NaN,0.0021,NaN,112.317216,1,Imaging,2010


Vamos a crear 3 dataframes (planetas, estrella y sistema) para trabajar con más facilidad

In [17]:
pl_df = exoplanets[['rowid','pl_name','pl_rade','pl_bmasse','pl_eqt','pl_orbper','pl_orbsmax','pl_insol']]

In [21]:
st_df = exoplanets[['rowid','hostname','st_teff','st_rad','st_mass','st_met','st_spectype','st_age','st_rotp']]

In [19]:
sy_df = exoplanets[['rowid','sy_dist','sy_pnum','discoverymethod','disc_year']]

Guardamos los dataframes en csvs para tener los datos limpios listo para usar.

In [22]:
pl_df.to_csv('../data/dataclean_pl.csv', sep=',', index=False)
st_df.to_csv('../data/dataclean_st.csv', sep=',', index=False)
sy_df.to_csv('../data/dataclean_sy.csv', sep=',', index=False)